### PR TITLE
[WIP] sp-profiler

### PR DESCRIPTION
**Pull request by mattrutherford**
WIP for a Profiler that can work inside WASM. Definitely want to emphasise the POC status at the moment to try and prove usefulness - will need decisions made on what the scope of this should be, as it has some conflicting ideas at the moment. Still quite some way from being in a state to merge, but I wanted to open a PR to begin discussion.

Currently supports modes for running benchmarks without the client, by output to stdout or a file.

Also supports output to tracing, if also configured in client.

The design is awkward due to the constraints of the situation, but perhaps it's useful enough to warrant this.

It can support tracing anywhere in wasm, using a very limited subset of the tracing API. There are certain scenarios where a span exit may be lost (due to fn returning early) - in this case we mark the span as bad (`sp_profiler_ok = false`) and send it on receipt of next span exit.

Trialing the recording of values for spans, which is awkward due to the requirement for identifiers to be const. 

A current POC of somewhere this might be useful is capturing the weight (along with execution time) of an extrinsic:

Screenshot 2020-04-24 at 17 00 38

### CODE REVIEW

File: primitives/profiler/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+use std::thread::JoinHandle;
+use std::sync::atomic::{AtomicU64, Ordering};
+use parking_lot::Mutex;
+use std::collections::{HashMap, BTreeMap};
+use tracing::{span, field, Level};
+
+const MAX_SPAN_STACK_LEN: usize = 100;
+
+///
+pub trait Profiler: Send {
```
Comment (bkchr) [inline primitives/profiler/src/lib.rs:13]: Why is that not in `sp-tracing`?

File: primitives/io/src/lib.rs [hunk 2]
```diff
@@ -63,6 +63,25 @@ mod batch_verifier;
...
 
+#[cfg(feature = "std")]
+#[macro_use]
+extern crate lazy_static;
+#[cfg(feature = "std")]
+use parking_lot::Mutex;
+#[cfg(feature = "std")]
+use sp_profiler::{Profiler, BasicProfiler, FileProfiler, TracingProfiler};
+#[cfg(feature = "std")]
+lazy_static! {
```
Comment (bkchr) [inline primitives/io/src/lib.rs:74]: A big no to this.

File: primitives/io/src/lib.rs [hunk 3]
```diff
@@ -123,7 +142,10 @@ pub trait Storage {
 	///
 	/// Returns the SCALE encoded hash.
 	fn root(&mut self) -> Vec<u8> {
-		self.storage_root()
+		let id = PROFILER.lock().create_span(module_path!(), "root");
+		let res = self.storage_root();
+		PROFILER.lock().exit_span(id);
```
Comment (bkchr) [inline primitives/io/src/lib.rs:147]: Why? These calls are already tracked automatically.

File: primitives/io/src/lib.rs [hunk 4]
```diff
@@ -842,6 +865,28 @@ pub trait Logging {
 	}
 }
 
+/// Interface that provides functions for profiling the runtime.
+#[runtime_interface]
+pub trait Profiling {
+	/// For enter_span to work correctly with the Tracing profiler,
+	/// the span `target` and `name` must also be registered in `sp_profiler::TracingProfiler`
+	fn enter_span(target: &str, name: &str) -> u64 {
+		PROFILER.lock().create_span(target, name)
```
Comment (bkchr) [inline primitives/io/src/lib.rs:874]: Why don't we forward this to the normal tracer?

File: primitives/io/src/lib.rs (hunk 2 above)
Comment (mattrutherford) [reply primitives/io/src/lib.rs:74]: I know this isn't the way to do it, but wanted a way to support profiling in situations where we don't have the client - eg. for isolated benchmarking of the runtime. Will figure out in due course

File: primitives/io/src/lib.rs (hunk 3 above)
Comment (mattrutherford) [reply primitives/io/src/lib.rs:147]: yes, this is left over from before and doesn't have to be there, will remove

File: primitives/io/src/lib.rs (hunk 4 above)
Comment (mattrutherford) [reply primitives/io/src/lib.rs:874]: For 2 reasons it's not that simple, first being that tracing has a requirement that span name and target be `const`. Second is that I wanted to support profiling when the client tracer is not available. Happy to make a compromise though and split things up if needed

File: primitives/io/src/lib.rs (hunk 2 above)
Comment (bkchr) [reply primitives/io/src/lib.rs:74]: The solution is to instantiate the tracer if you need it.

**Comment by mattrutherford** [comment]
Thanks for the review and I've decided to close this in favour of a new approach.

Will move the span registration code into sp-tracing and remove the file-based and stdout profiler, instead utilising tracing system at all times. I will also document how to use this for benchmark scenarios.

I think that moving value recording into tracing events makes most sense in a WASM context, since we're not restricted to const identifiers in that case.

Will open a new PR.

**Pull request closed without merging**